log/loki.go

```diff
@@ -0,0 +1,419 @@
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				// TODO filter the additional ones
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// TODO have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case <-ticker.C:
+			ch := make(chan struct{})
+			pushCh <- ch
+			<-ch
+		}
+	}
+}
+
+func (h *lokiHook) push(msgs []tmpMsg, dropped int) (int, error) {
+	if len(msgs) == 0 {
+		return 0, nil
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	// We can technically cutoff during the addMsg phase, which will be even better if we sort after
+	// it as well ... maybe
+	cutOff := time.Now().Add(-time.Millisecond * 500).UnixNano() // probably better to be configurable
+
+	cutOffIndex := sort.Search(len(msgs), func(i int) bool {
+		return !(msgs[i].t < cutOff)
+	})
+
+	if cutOffIndex > len(msgs) {
+		cutOffIndex = len(msgs)
+	}
+
+	strms := new(lokiPushMessage)
+
+	for _, msg := range msgs[:cutOffIndex] {
+		strms.add(msg)
+	}
+	if dropped != 0 {
+		labels := make(map[string]string, 2+len(h.labels))
+		labels["level"] = logrus.WarnLevel.String()
+		labels["dropped"] = strconv.Itoa(dropped)
+		for _, params := range h.labels {
+			labels[params[0]] = params[1]
+		}
+
+		msg := tmpMsg{
+			labels: labels,
+			msg: fmt.Sprintf("k6 dropped some packages because they were above the limit of %d/%s",
+				h.limit, h.pushPeriod),
+			t: msgs[cutOffIndex-1].t,
+		}
+		strms.add(msg)
+	}
+
+	var b bytes.Buffer
+	_, err := strms.WriteTo(&b)
+	if err != nil {
+		return cutOffIndex, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", h.addr, &b)
+	if err != nil {
+		return cutOffIndex, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := h.client.Do(req)
+	if res != nil {
+		_, _ = io.Copy(ioutil.Discard, res.Body)
+		_ = res.Body.Close()
+	}
+	return cutOffIndex, err
+}
+
+func mapEqual(a, b map[string]string) bool {
```

Comment:
Move this to `lib` (ugh) or a util/helpers package?

Reply:
it's not used anywhere else ... I hate moving things to other packages in anticipation that they will be reused ...

Reply:
Sure, but in this case there's nothing specific to `log` about it and it could be useful elsewhere. Leaving it here ensures it will never be reused. :) Not a big deal, though.

Reply:
Technically, something almost like this is used in https://github.com/loadimpact/k6/blob/0e94653c9954c17b2b1a480b456e23ec120feff2/stats/stats.go#L192-L206

That said, I don't mind having this here, it's unfortunately "the Go way", until we start having some generics :sob: